File: 03_Factory_Method/cpp/src/main.cpp
#include <iostream>
#include <memory>

// 03_Factory_Method Pattern - C++ Implementation

// Product
class Document{
    public:
        virtual ~Document() = default;
        virtual void open() const = 0;
        virtual void edit() const = 0;
        virtual void save() const = 0;
};

class WordDocument : public Document{
    public:
        void open() const override {
            std::cout << "Word: open\n";
        }

        void edit() const override {
            std::cout << "Word: edit\n";
        }

        void save() const override {
            std::cout << "Word: save\n";
        }
};

class PdfDocument : public Document{
    public:
        void open() const override {
            std::cout << "Pdf: open\n";
        }

        void edit() const override {
            std::cout << "Pdf: edit\n";
        }

        void save() const override {
            std::cout << "Pdf: save\n";
        }
};

// Creator
class DocumentEditor {
    public:
        virtual ~DocumentEditor() = default;

        // Factory method
        virtual std::unique_ptr<Document> createDocument() const = 0;

        // Template methods
        void run() const {
            auto doc = createDocument();
            doc -> open();
            doc->edit();
            doc->save();
        }
};

class WordEditor : public DocumentEditor {
    public:
        std::unique_ptr<Document> createDocument() const override {
            return  std::make_unique<WordDocument>();
        }
};

class PdfEditor : public DocumentEditor {
    public:
        std::unique_ptr<Document> createDocument () const override {
            return std::make_unique<PdfDocument>();
        }
};

int main() {
    std::cout << "=== [C++] 03_Factory_Method ===" << std::endl;
    
    std::unique_ptr<DocumentEditor> editor = std::make_unique<WordEditor>();

    editor->run();

    return 0;
}

File: 03_Factory_Method/rs/src/main.rs
trait Document {
    fn open(&self);
    fn edit(&self);
    fn save(&self);
}

struct WordDocument;
impl Document for WordDocument {
    fn open(&self) {
        println!("Word: open");
    }
    
    fn edit(&self) {
        println!("Word: edit");
    }
    
    fn save(&self) {
        println!("Word: save");        
    }
}

struct PdfDocument;
impl Document for PdfDocument {
    fn open(&self) {
        println!("Pdf: open");
    }

    fn edit(&self) {
        println!("Pdf: edit");
    }

    fn save(&self) {
        println!("Pdf: save");
    }
}

// 2. CREATOR (Abstract)
trait Editor {
    type Doc: Document;
    // ===> THIS IS THE FACTORY METHOD <===
    fn create_document(&self) -> Self::Doc;
    // Core business logic relies on the factory method:
    fn run(&self) {
        let document = self.create_document();
        document.open();
        document.edit();
        document.save();
    }
}
// 3. CONCRETE CREATORS (They decide which Product to instantiate)
struct WordEditor;
impl Editor for WordEditor {
    type Doc = WordDocument;
    fn create_document(&self) -> Self::Doc {
        WordDocument
    }
}
struct PdfEditor;
impl Editor for PdfEditor {
    type Doc = PdfDocument;
    fn create_document(&self) -> Self::Doc {
        PdfDocument
    }
}


// Method 3

fn run_editor<F, D>(create: F) 
where  
    F: Fn() -> D,
    D: Document,
{
    let doc = create();
    doc.open();
    doc.edit();
    doc.save();
}


fn main() {
    println!("=== [Rust] 03_Factory_Method ===");
    
    // Client works with the Creator:
    let pdf_editor = PdfEditor;
    let word_editor = WordEditor;

    pdf_editor.run();
    word_editor.run();

    run_editor(|| WordDocument);
    run_editor(|| PdfDocument);
}
